micro-rdk/src/common/data_store.rs

```diff
@@ -0,0 +1,475 @@
+
+    /// Resets the data store.
+    fn clear(&mut self);
+}
+
+/// StaticMemoryDataStore is an entity that governs the static bytes memory
+/// reserved in DATA_STORE. The memory is segmented based according to the DataCollectors expected
+/// (identified by collector keys) and each segment view is treated as a separate ring buffer of SensorData
+/// messages. Currently, an equal amount of space is alloted to each collector, which will affect
+/// the maximum allowed size of a single message (computed as the length of DATA_STORE divided by
+/// the number of collector keys). It should be treated as a global struct that should only be initialized once
+/// and is not thread-safe (all interactions should be blocking).
+pub struct StaticMemoryDataStore {
+    buffers: Vec<LocalRb<u8, &'static mut [MaybeUninit<u8>]>>,
+    collector_keys: Vec<ResourceMethodKey>,
+}
+
+impl StaticMemoryDataStore {
+    pub fn new(collector_keys: Vec<ResourceMethodKey>) -> Result<Self, DataStoreError> {
+        if !DATA_STORE_INITIALIZED.fetch_or(true, Ordering::SeqCst) {
+            let len_per_buffer = unsafe { DATA_STORE.len() } / collector_keys.len();
+            let mut buffers = Vec::new();
+            for i in 0..(collector_keys.len()) {
+                let start_idx = i * len_per_buffer;
+                let end_idx = (i + 1) * len_per_buffer;
+                unsafe {
+                    buffers.push(LocalRb::from_raw_parts(
+                        &mut DATA_STORE[start_idx..end_idx],
+                        0,
+                        0,
+                    ));
+                }
+            }
+            return Ok(Self {
+                buffers,
+                collector_keys,
+            });
+        }
+        Err(DataStoreError::DataStoreInitialized)
+    }
+
+    fn get_index_for_collector(
+        &self,
+        collector_key: &ResourceMethodKey,
+    ) -> Result<usize, DataStoreError> {
+        self.collector_keys
+            .iter()
+            .position(|key| key == collector_key)
+            .ok_or(DataStoreError::UnknownCollectorKey(collector_key.clone()))
+    }
+}
+
+impl DataStore for StaticMemoryDataStore {
+    fn store_message(
+        &mut self,
+        collector_key: ResourceMethodKey,
+        message: SensorData,
+    ) -> Result<(), DataStoreError> {
+        let buffer_index = self.get_index_for_collector(&collector_key)?;
+        let buffer = &mut self.buffers[buffer_index];
+        let encode_len = message.encoded_len();
+        if encode_len + 5 > buffer.vacant_len() {
+            return Err(DataStoreError::DataBufferFull(collector_key, message));
+        }
+        buffer
+            .push(0)
+            .map_err(|_| DataStoreError::DataWriteFailure)?;
+        let len_bytes = (encode_len as u32).to_le_bytes();
+        buffer.push_slice(&len_bytes);
+
+        unsafe {
+            let mut prod = Producer::new(buffer as &LocalRb<u8, &'static mut [MaybeUninit<u8>]>);
```

Comment:
- I've perhaps misunderstood, but how would I use write_all to prevent a separate allocation that I would have to copy the bytes into? This was why I didn't use `buffer.push_slice()`
- I find it maybe equally awkward to store something on the struct that contains a reference to something else on the struct and the pattern of creating a Producer on the fly is one that the ringbuf library uses internally in implementing `Rb`

Reply:
You are right, I got myself tangled up in the traits. It's unfortunate, because Prost's `encode` only operates on `BytesMut`, and `Producer` can write to an `io::Writer` and `BytesMut` offers `.writer` to produce a `Writer`, but as far as I can see you can't get a `Producer` to appear as a `BytesMut` such that you can give it to `Prost::encode` nor can you get `Prost::encode` to write to an `io::Writer` which would be another way, one of which was what I was hoping for and convinced myself worked.